src/SingleVault.sol

```diff
@@ -1,99 +1,64 @@
+import {ISingleVault} from "src/interface/ISingleVault.sol";
 
-contract SingleVault is ISingleVault, ERC4626Upgradeable, TimelockControllerUpgradeable, ReentrancyGuardUpgradeable {
-    using Math for uint256;
+/* ynETH Pre-Launch Vault */
 
-    bytes32 private constant ERC4626StorageLocation = 0x0773e532dfede91f04b12a73d3d2acd361424f41f76b4fb79f090161e36b4e00;
-    IStakeManager private constant stakeManager = IStakeManager(0x1adB950d8bB3dA4bE104211D5AB038628e477fE6);
+contract SingleVault is ISingleVault, ERC4626Upgradeable, AccessControlUpgradeable, ReentrancyGuardUpgradeable {
+    using Math for uint256;
 
     constructor() {
         _disableInitializers();
     }
 
-    /**
-     * @dev Initializes the SingleVault contract.
-     * @param asset_ The address of the ERC20 asset.
-     * @param name_ The name of the vault.
-     * @param symbol_ The symbol of the vault.
-     * @param admin_ The address of the admin.
-     * @param minDelay_ The minimum delay for timelock.
-     * @param proposers_ The addresses of the proposers.
-     * @param executors_ The addresses of the executors.
-     */
-    function initialize(
-        IERC20 asset_,
-        string memory name_,
-        string memory symbol_,
-        address admin_,
-        uint256 minDelay_,
-        address[] calldata proposers_,
-        address[] calldata executors_
-    ) public initializer {
-        _verifyParamsAreValid(asset_, name_, symbol_, admin_, proposers_, executors_);
-        __TimelockController_init(minDelay_, proposers_, executors_, admin_);
+    function initialize(IERC20 asset_, string memory name_, string memory symbol_, address admin_) public initializer {
+        _verifyParamsAreValid(asset_, name_, symbol_, admin_);
         __ERC20_init(name_, symbol_);
         __ERC4626_init(asset_);
         __AccessControl_init();
         __ReentrancyGuard_init();
         _grantRole(DEFAULT_ADMIN_ROLE, admin_);
     }
 
-    function _verifyParamsAreValid(
-        IERC20 asset_,
-        string memory name_,
-        string memory symbol_,
-        address admin_,
-        address[] calldata proposers_,
-        address[] calldata executors_
-    ) internal pure {
-        if (asset_ == IERC20(address(0))) revert AssetZeroAddress();
-        if (bytes(name_).length == 0) revert NameEmpty();
-        if (bytes(symbol_).length == 0) revert SymbolEmpty();
-        if (admin_ == address(0)) revert AdminZeroAddress();
-        if (proposers_.length == 0) revert ProposersEmpty();
-        if (executors_.length == 0) revert ExecutorsEmpty();
+    receive() external payable nonReentrant {
+        if (msg.value > 0) {
+            _mintSharesForETH(msg.value);
+        }
     }
 
-    function _retrieveERC4626Storage() internal pure returns (ERC4626Storage storage $) {
-        assembly {
-            $.slot := ERC4626StorageLocation
+    fallback() external payable nonReentrant {
+        if (msg.value > 0) {
+            _mintSharesForETH(msg.value);
```

Comment:
this implementation assumes the rate of ETH:Shares is 1:1

Which it may not be even for the starter vault.

For example, if i send WETH directly to the vault, it will increase the rate, as ERC4626Upgradeable counts totalAssets using `$._asset.balanceOf(address(this))`

the easiest implementation here is probably using the _mint function, and previewDeposit.

Have to be careful to mint WETH *only after* calling previewDeposit to not inflate totalAssets before it's processed